File: src/main/java/com/gluck/jobtracker/service/UserDetailsServiceImpl.java
package com.gluck.jobtracker.service;

import com.gluck.jobtracker.repository.UserEntity;
import com.gluck.jobtracker.repository.UserRepository;
import lombok.AllArgsConstructor;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.Collections;

@AllArgsConstructor
@Service
public class UserDetailsServiceImpl implements UserDetailsService {

    private final UserRepository repository;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        String safeUsername = username;

        UserEntity userEntity = repository.findByUsername(safeUsername);
        if (userEntity == null) {
            throw new UsernameNotFoundException("User by " + safeUsername + " not found");
        }

        return User.withUsername(userEntity.getUsername())
            .password(userEntity.getPassword())
            .authorities(Collections.singletonList(new SimpleGrantedAuthority(userEntity.getRole())))
            .build();
    }
}

File: src/main/kotlin/com/gluck/jobtracker/service/UserDetailsServiceImpl.kt
package com.gluck.jobtracker.service

import com.gluck.jobtracker.repository.UserRepository
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

@Service
class UserDetailsServiceImpl(private val repository: UserRepository): UserDetailsService {
    override fun loadUserByUsername(username: String): UserDetails {

        val safeUsername = username // ?: throw UsernameNotFoundException("Username is null")

        val userEntity = repository.findByUsername(safeUsername)
            ?: throw UsernameNotFoundException("User by $safeUsername not found")

        return User.withUsername(userEntity.username)
            .password(userEntity.password)
            .authorities(SimpleGrantedAuthority(userEntity.role))
            .build()

    }
}